### ISSUE TITLE
Add support for Fallout 4 from the Epic Games Store

### ISSUE THREAD

**Issue by Ortham** [loot/loot#1965]
Fallout 4 is now [available](https://store.epicgames.com/en-US/p/fallout-4) on the Epic Games Store. For some reason even though the game is almost a decade old, that's just the base game: the GOTY edition that includes all the DLC is [here](https://store.epicgames.com/en-US/p/fallout-4--game-of-the-year-edition). They're currently 40% off until the 5th of May, at ~ £10 and ~ £21 respectively.

**Comment by Ortham** [comment]
I bought and downloaded the regular edition of the game, some details:

- The game would get stuck on a waiting screen before the main menu until I allowed it to link to my Epic account (it opens a page in your default web browser).
- The game does have Bethesda.net integration (i.e. a Mods menu), I didn't log in though.
- The `My Games` and `AppData\Local` folders are both named `Fallout4 EPIC`.
- The game reads `Plugins.txt` (with that capitalisation).
- The game's EGS `AppName` is `61d52ce4d09d41e48800c22784d13ae8`
- `Fallout4.ccc` exists same as for other stores.
- The plugin header version is `1`:
```
> fs.readdirSync(path).filter(p => p.endsWith('.esp') || p.endsWith('.esm') || p.endsWith('.esl')).map(p => { let b = Buffer.alloc(34); let h = fs.openSync(`${path}/${p}`); fs.readSync(h, b, 0, 0x1E); fs.readSync(h, b, 0, 4); return [p, b.readFloatLE()]; })
[
  [ 'ccBGSFO4044-HellfirePowerArmor.esl', 1 ],
  [ 'ccBGSFO4046-TesCan.esl', 1 ],
  [ 'ccBGSFO4096-AS_Enclave.esl', 1 ],
  [ 'ccBGSFO4110-WS_Enclave.esl', 1 ],
  [ 'ccBGSFO4115-X02.esl', 1 ],
  [ 'ccBGSFO4116-HeavyFlamer.esl', 1 ],
  [ 'ccFSVFO4007-Halloween.esl', 1 ],
  [ 'ccOTMFO4001-Remnants.esl', 1 ],
  [ 'ccSBJFO4003-Grenade.esl', 1 ],
  [ 'Fallout4.esm', 1 ]
```
- The BA2 header version varies between `1`, `7` and `8`:
```
> path = 'D:\\Games\\Epic Games Store\\molossusPublic\\Data'
'D:\\Games\\Epic Games Store\\molossusPublic\\Data'
> fs = require('fs')
> fs.readdirSync(path).filter(p => p.endsWith('.ba2')).map(p => { let b = Buffer.alloc(4); let h = fs.openSync(`${path}/${p}`); fs.readSync(h, b, 0, 4); fs.readSync(h, b, 0, 4); return [p, b.readUint32LE()]; })
[
  [ 'ccBGSFO4001-PipBoy(Black) - Main.ba2', 1 ],
  [ 'ccBGSFO4001-PipBoy(Black) - Textures.ba2', 7 ],
  [ 'ccBGSFO4004-PipBoy(Camo02) - Main.ba2', 1 ],
  [ 'ccBGSFO4004-PipBoy(Camo02) - Textures.ba2', 7 ],
  [ 'ccBGSFO4006-PipBoy(Chrome) - Main.ba2', 1 ],
  [ 'ccBGSFO4006-PipBoy(Chrome) - Textures.ba2', 7 ],
  [ 'ccBGSFO4016-Prey - Main.ba2', 1 ],
  [ 'ccBGSFO4016-Prey - Textures.ba2', 7 ],
  [ 'ccBGSFO4018-GaussRiflePrototype - Main.ba2', 8 ],
  [ 'ccBGSFO4018-GaussRiflePrototype - Textures.ba2', 8 ],
  [ 'ccBGSFO4019-ChineseStealthArmor - Main.ba2', 1 ],
  [ 'ccBGSFO4019-ChineseStealthArmor - Textures.ba2', 7 ],
  [ 'ccBGSFO4020-PowerArmorSkin(Black) - Main.ba2', 8 ],
  [ 'ccBGSFO4020-PowerArmorSkin(Black) - Textures.ba2', 8 ],
  [ 'ccBGSFO4038-HorseArmor - Main.ba2', 1 ],
  [ 'ccBGSFO4038-HorseArmor - Textures.ba2', 7 ],
  [ 'ccBGSFO4044-HellfirePowerArmor - Main.ba2', 8 ],
  [ 'ccBGSFO4044-HellfirePowerArmor - Textures.ba2', 8 ],
  [ 'ccBGSFO4046-TesCan - Main.ba2', 8 ],
  [ 'ccBGSFO4046-TesCan - Textures.ba2', 8 ],
  [ 'ccBGSFO4096-AS_Enclave - Main.ba2', 8 ],
  [ 'ccBGSFO4096-AS_Enclave - Textures.ba2', 8 ],
  [ 'ccBGSFO4110-WS_Enclave - Main.ba2', 8 ],
  [ 'ccBGSFO4110-WS_Enclave - Textures.ba2', 8 ],
  [ 'ccBGSFO4115-X02 - Main.ba2', 8 ],
  [ 'ccBGSFO4115-X02 - Textures.ba2', 8 ],
  [ 'ccBGSFO4116-HeavyFlamer - Main.ba2', 8 ],
  [ 'ccBGSFO4116-HeavyFlamer - Textures.ba2', 8 ],
  [ 'ccFRSFO4001-HandmadeShotgun - Main.ba2', 1 ],
  [ 'ccFRSFO4001-HandmadeShotgun - Textures.ba2', 7 ],
  [ 'ccFSVFO4001-ModularMilitaryBackpack - Main.ba2', 1 ],
  [ 'ccFSVFO4001-ModularMilitaryBackpack - Textures.ba2', 7 ],
  [ 'ccFSVFO4002-MidCenturyModern - Main.ba2', 8 ],
  [ 'ccFSVFO4002-MidCenturyModern - Textures.ba2', 8 ],
  [ 'ccFSVFO4007-Halloween - Main.ba2', 8 ],
  [ 'ccFSVFO4007-Halloween - Textures.ba2', 8 ],
  [ 'ccOTMFO4001-Remnants - Main.ba2', 8 ],
  [ 'ccOTMFO4001-Remnants - Textures.ba2', 8 ],
  [ 'ccSBJFO4003-Grenade - Main.ba2', 8 ],
  [ 'ccSBJFO4003-Grenade - Textures.ba2', 8 ],
  [ 'Fallout4 - Animations.ba2', 1 ],
  [ 'Fallout4 - Interface.ba2', 8 ],
  [ 'Fallout4 - Localization.ba2', 8 ],
  [ 'Fallout4 - Materials.ba2', 8 ],
  [ 'Fallout4 - Meshes.ba2', 1 ],
  [ 'Fallout4 - MeshesExtra.ba2', 1 ],
  [ 'Fallout4 - Misc - Beta.ba2', 1 ],
  [ 'Fallout4 - Misc - Debug.ba2', 1 ],
  [ 'Fallout4 - Misc.ba2', 8 ],
  [ 'Fallout4 - Nvflex.ba2', 1 ],
  [ 'Fallout4 - Nvtess.ba2', 1 ],
  [ 'Fallout4 - Shaders.ba2', 8 ],
  [ 'Fallout4 - Sounds.ba2', 1 ],
  [ 'Fallout4 - Startup.ba2', 8 ],
  [ 'Fallout4 - Textures1.ba2', 7 ],
  [ 'Fallout4 - Textures2.ba2', 7 ],
  [ 'Fallout4 - Textures3.ba2', 7 ],
  [ 'Fallout4 - Textures4.ba2', 7 ],
  [ 'Fallout4 - Textures5.ba2', 7 ],
  [ 'Fallout4 - Textures6.ba2', 7 ],
  [ 'Fallout4 - Textures7.ba2', 7 ],
  [ 'Fallout4 - Textures8.ba2', 7 ],
  [ 'Fallout4 - Textures9.ba2', 7 ],
  [ 'Fallout4 - Voices.ba2', 1 ],
  [ 'Fallout4 - Voices_de.ba2', 1 ],
  [ 'Fallout4 - Voices_es.ba2', 1 ],
  [ 'Fallout4 - Voices_fr.ba2', 1 ],
  [ 'Fallout4 - Voices_it.ba2', 1 ],
  [ 'Fallout4 - Voices_jp.ba2', 1 ],
  [ 'Fallout4 - Voices_rep_ja.ba2', 1 ]
]
```

**Comment by Ortham** [comment]
I've done the necessary changes as of c2a991eb45e9d3d6d511f10fe0840664a1aaa9a0, loot/libloot@8a02621bb2b7b8834627808e170694cef47b90e7 and Ortham/libloadorder@5684a78ca6b18316947d4ed6df3ed073caa678fb. The latter aren't released yet (and so LOOT hasn't been updated to use them) because I'm waiting on any news on Starfield modding support.

**Comment by pStyl3** [comment]
>.. because I'm waiting on any news on Starfield modding support.

The next Starfield Update is in Beta and will be released to the public later in May:
[Details](https://www.reddit.com/r/Starfield/comments/1choql8/starfield_update_11133_may_1_2024_beta/) - [Video](https://www.youtube.com/watch?v=3ObHRMHtTMY)

While this update seems to improve on a lot of things, modding support doesn't seem to be a part of it.

At the end of the video they say:

> Jess Finster:
.. is there anything else that you wanna talk about, that's down the horizon?
> 
> Tim Lamb:
So, the team is working on our first land vehicle. So you can drive around planets.
Obviously we are working on more quests and the massive Shattered Space expansion.
> 
> Jess Finster:
That's all very exciting. What about creations, anything you can share there?
> 
> Tim Lamb:
I think we mentioned it earlier this year, but we've send out our Beta toolkit to verified creators, so hopefully more information on that soon.

**Comment by Ortham** [comment]
Yikes, after promising updates every 6 weeks and not sharing anything since mid-March, I thought they were holding back all news so they could do a big launch announcement of the modding tools this week, but instead their most prominent heading is that we're going to be getting proper local maps in two weeks...

Well, the pathetically useless local maps that the game launched with were one of the reasons that I bounced off the game hard, so it's good they're finally being addressed, but they declined to share any info (e.g. an ETA) on the modding tools and that doesn't fill me with confidence. I can't see them landing until at least late June, and honestly wouldn't be surprised if Bethesda just decide to release them on the game's anniversary as that would only be two six-week cycles away at that point, and making a big announcement over how much the game has improved over its first year plus the release of modding tools seems like better marketing than releasing months "late" (according to general "up to 6 months" expectations set by previous titles).

Given the uncertainty I don't think it's worth holding back the breaking changes I've already got, if I have to make more down the line then so be it.

(Having said all that, they will of course release the day after I release new major versions of my stuff.)

Thanks for sharing @pStyl3!
